Support "is None" constraints from if statements during inference

david-yz-liu opened the pull request:
<!--

Thank you for submitting a PR to astroid!

To ease our work reviewing your PR, do make sure to mark the complete the following boxes.

-->

## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

Alright, this is a big one so please don't hold back. This PR adds infrastructure to support *constraints* on values during inference. Here's a simple example:

```python
from astroid import extract_node

node = extract_node("""
def f(b):
    if b:
        x = 1
    else:
        x = None
    
    if x is not None:
        x #@
""")
print(node.inferred())
```

This prints `[<Const.int l.4 at 0x14232a6e6d0>, <Const.NoneType l.6 at 0x142329b38b0>]`, using both possible values from the assignment statements. Of course, because of the `x is not None` if condition, only the int is possible, but astroid currently doesn't take this condition into account.

This PR changes how inference works so that only `[<Const.int l.4 at 0x14232a6e6d0>]` is printed.

### Scope

I reviewed the existing pylint issues related to this (they are generally labelled as "topic-control-flow"). There were four types of constraints that were brought up:

1. `x is None`/`x is not None`
2. `x`/`not x` (truthy/falsey value)
3. `isinstance(x, ...)`
4. `x == ...`
5. Combinations of the above with `and`/`or`

There were also three primary locations where these constraints could be inferred (orthogonal to the constraints themselves):

1. if statement (as in above example)
2. `and`/`or` clauses. For example, `x is not None and x[0] == 1`; the second `x` should not be inferred as `None`
3. ternary if expressions. For example, `x[0] if x is not None else 1`; the first `x` should not be inferred as `None`
4. reassignment within an if statement, e.g.

    ```python
    def f():
        a = None
        if a is None:
            a = 1
        return -a
    ```
5. early return/raises within an if statement.
6. assert statements

This pull request addresses just the item 1's in each list. I used an approach that can handle 1-5 of the constraint types, and 1-3 of the locations. If this approach is approved, I think it would be pretty straightforward to make additional PRs for these items. (More on the other constraint locations below.) 

### Implementation overview

I added a new `Constraint` class to represent a constraint on a variable, like `x is not None`.
- This is an abstract class, and can be subclassed for different constraint types.

When a variable is inferred (see `inference.py`, `infer_name`), we search the ancestor chain from the node and look for containing if statements, using these to accumulate constraints on the variable. These constraints are added to the inference context (added a new attribute to `InferenceContext`).

Then, in `_infer_stmts` (`bases.py`), we filter the inferred values based on whether or not they satisfy all constraints for that variable. 
- `Uninferable` is considered to satisfy all constraints.
- ⚠️ If there is at least one inferred value, but *all* inferred values are filtered out, a new `Uninferable` value is yielded, rather than yielding nothing at all. I think this is the conservative approach, to avoid errors from the `raise_if_nothing_inferred` decorator.

The above also applies to attributes (see `infer_attribute`).

### Some comments/questions

1. Because constraints are gathered by searching up the ancestor chain, this PR doesn't handle real control flow issues, e.g. with the reassignment example from above.
2. Because the filtering occurs in `_infer_stmts`, it only applies to variables and attributes. So for example, subscript nodes (e.g. `my_array[0]`) won't have their inferred values filtered.
3. Gathering constraints occurs every time `infer_name`/`infer_attribute` is called, and since this requires going up the ancestor tree, this adds some overhead to inference. In principle constraints can be gathered through an AST visitor, but I thought that would be more complex so didn't choose it. (Thoughts welcome of course!)
4. I think adding the constraints to `InferenceContext` is correct, but to be honest I'm fuzzy on the nuances of this class and how it's used.
5. Do you like the name "constraint" here? I go back on forth on it, but haven't come up with anything better.

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Related Issue

<!--
If this PR fixes a particular issue, use the following to automatically close that issue
once this PR gets merged:

Closes #XXX
-->

Closes #791
Closes PyCQA/pylint#157
Closes PyCQA/pylint#1472
Closes PyCQA/pylint#2016
Closes PyCQA/pylint#2631
Closes PyCQA/pylint#2880

DanielNoord commented:
Hi @david-yz-liu, thanks for all of the work here. I think the size (and importance) of this PR has put off people from reviewing this but I think we should try and start the process of getting this merged.

The first question I would have before doing a full review is how we could expand this to include conditional reassignments. I think it would be every important to support the following code pattern:
```python
def func():
    if True:
        to_infer = 1
    else:
        to_infer = 2

    print(to_infer)
```

This does not need to be added in this PR, but we should at least have a general idea of how we could implement this within the same system. Do/did you have any ideas about this yourself?

david-yz-liu commented:
@DanielNoord wow, thank you! So sorry for the delay, I think I saw your message and meant to reply months ago but then... forgot. My bad!

Your question is a great one. It's been a while since I've thought about this, but I'll try to give a decent answer now (and perhaps a more thorough one later). I believe my current approach piggybacks off of the way `InferenceContext`s are passed around when doing inference, so the general rule for what this PR can or can't do is "we can accumulate Contraints whenever we can accumulate other things in `InferenceContext`s". The examples I gave in the PR description are places where a context is passed from a parent down to a child node in the AST, so they're pretty straightforward.

*However* as far I as can remember, existing `InferenceContexts` don't get passed across statements in a block. I think this is fundamentally why pylint's [`VariablesChecker`](https://github.com/PyCQA/pylint/blob/main/pylint/checkers/variables.py) is as complex as it is?

But anyway to answer your question, I think that this PR _can_ be extended to handle the example you provided, but it wouldn't be by adding to my code very much, but rather by changing how these `InferenceContext`s (or some other "context" state) is passed around, using techniques similar to what's happening in the `VariablesChecker`. Or, using control flow graphs (😅) to precisely model the possible execution paths, and then accumulating Contraints across paths in the graph.

Please let me know if that answers your question, or if there are any follow-ups! Happy to be of service :)

DanielNoord commented:
> @DanielNoord wow, thank you! So sorry for the delay, I think I saw your message and meant to reply months ago but then... forgot. My bad!
> 
> Your question is a great one. It's been a while since I've thought about this, but I'll try to give a decent answer now (and perhaps a more thorough one later). I believe my current approach piggybacks off of the way `InferenceContext`s are passed around when doing inference, so the general rule for what this PR can or can't do is "we can accumulate Contraints whenever we can accumulate other things in `InferenceContext`s". The examples I gave in the PR description are places where a context is passed from a parent down to a child node in the AST, so they're pretty straightforward.
> 
> _However_ as far I as can remember, existing `InferenceContexts` don't get passed across statements in a block. I think this is fundamentally why pylint's [`VariablesChecker`](https://github.com/PyCQA/pylint/blob/main/pylint/checkers/variables.py) is as complex as it is?
> 
> But anyway to answer your question, I think that this PR _can_ be extended to handle the example you provided, but it wouldn't be by adding to my code very much, but rather by changing how these `InferenceContext`s (or some other "context" state) is passed around, using techniques similar to what's happening in the `VariablesChecker`. Or, using control flow graphs (😅) to precisely model the possible execution paths, and then accumulating Contraints across paths in the graph.
> 
> Please let me know if that answers your question, or if there are any follow-ups! Happy to be of service :)

I think that answers it pretty clearly. As you can see I spent some time familiarising myself with the code 😄 I think this could indeed be a very good first step towards some better control flow. I have been thinking about doing something with `frames` and line numbering to do some more difficult constraints, but at the moment I don't see how we couldn't change the current approach to do that. It all seems flexible enough.

Let me know what you think of my changes. I made the tests pass on 3.7 and made some of the typing a little more strict. I appreciate that you had the constraints allow for other nodes than `If`, but for ease of later expansion I think it makes sense to keep it tighter now and expand the typing if needed later on.

david-yz-liu commented:
> Let me know what you think of my changes. I made the tests pass on 3.7 and made some of the typing a little more strict. I appreciate that you had the constraints allow for other nodes than `If`, but for ease of later expansion I think it makes sense to keep it tighter now and expand the typing if needed later on.

Will do. I'll try to look at this more closely tonight or tomorrow. 👍

Pierre-Sassoulas commented on ChangeLog:
```diff
@@ -6,12 +6,20 @@ What's New in astroid 2.12.0?
 =============================
 Release date: TBA
 
+
```
```suggestion
```

david-yz-liu commented:
@DanielNoord Okay I reviewed all of your changes, they look good to me! Yeah I was going for more general types but happy to stay strict for now and expand later if needed. Also thanks for making that Python 3.7 fix, that would have taken me a while to track down 😅.

Anyway as far as I can tell, this is good from my end, just waiting for more detailed review/feedback.

Pierre-Sassoulas commented on astroid/constraint.py:
```diff
@@ -0,0 +1,142 @@
+        if inferred is util.Uninferable:
+            return True
+
+        if self.negate and matches(inferred, nodes.Const(None)):
+            return False
+        if not self.negate and not matches(inferred, nodes.Const(None)):
+            return False
+
+        return True
```
```suggestion
        return bool(self.negate) ^ matches(inferred, nodes.Const(None))
```

```suggestion
        return bool(self.negate) ^ bool(matches(inferred, nodes.Const(None)))
```

I guess a xor would be slightly more efficient with a comparison less ? Not sure if we need the cast to bool on ``matches``.

david-yz-liu replied:
Thanks @Pierre-Sassoulas, I made a commit to use XOR. `self.negate` is a bool, and `matches` returns a `bool`, so we didn't need the explicit casting.

Pierre-Sassoulas reviewed (approved):
LGTM. The tests in particular are diversified and very meticulous. This will close a ton of issues and open a lot of possibilities in pylint too.

DanielNoord commented on astroid/constraint.py:
```diff
@@ -0,0 +1,137 @@
+# Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
```
Just thinking in terms of future expandability:
Should we make this a module with a `__init__.py` and put `get_constraints` in a `__all__`.

I can see these classes and functions change over time as we might expand the control flow inference and by putting things explicitly in `__all__` we don't need to worry about backwards incompatibility.

DanielNoord commented on astroid/constraint.py:
```diff
@@ -0,0 +1,137 @@
+    Constraints are computed statically by analysing the code surrounding expr.
+    Currently this only supports constraints generated from if conditions.
+    """
+    current_node: nodes.NodeNG = expr
```
```suggestion
    current_node = expr
```

This is probably left over from my previous changes. Forgot to remove it.

DanielNoord commented on astroid/constraint.py:
```diff
@@ -0,0 +1,137 @@
+    Currently this only supports constraints generated from if conditions.
+    """
+    current_node: nodes.NodeNG = expr
+    constraints = {}
```
```suggestion
    constraints: Dict[nodes.If, Constraint] = {}
```

DanielNoord commented:
@cdce8p Do you still want to review this before we merge this? I think with my last comments addressed this is almost good to go!

cdce8p commented:
> @cdce8p Do you still want to review this before we merge this? I think with my last comments addressed this is almost good to go!

If you're ok with waiting a bit longer, I could take a look in about 2-3 weeks. Will have much more time then to catch up on everything.

DanielNoord commented:
> > @cdce8p Do you still want to review this before we merge this? I think with my last comments addressed this is almost good to go!
> 
> If you're ok with waiting a bit longer, I could take a look in about 2-3 weeks. Will have much more time then to catch up on everything.

Fine with me! Should be around the time that I finish with `argparse` so I'll have time to assist with anything that comes up then 😄

cdce8p commented:
@DanielNoord Sounds good to me!

jacobtylerwalls reviewed (approved):
Looks great! I can help with merging main or resolving the last set of comments.

Pierre-Sassoulas reviewed (approved):
👌

david-yz-liu commented:
Sorry for the commit spam, I'm currently only running 3.10 locally. If this continues I'll fix it properly...

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                matches(left, node)
+                and matches(right, cls.CONST_NONE)
+                or matches(left, cls.CONST_NONE)
```
`cls.CONST_NONE` is a type error since `cls` is bound to `Constraint`. Here it can only ever be a `NoneConstraint` or a subclass of it. To reflect that, you need to create a new TypeVar unfortunately. At least until mypy supports the new `Self` typing. I.e.
```py
_NoneConstraintT = TypeVar("_NoneConstraintT", bound="NoneConstraint")
```

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+            left = expr.left
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                matches(left, node)
+                and matches(right, cls.CONST_NONE)
+                or matches(left, cls.CONST_NONE)
+                and matches(right, node)
```
Do we need to check both directions? How many people are going to write `None is x.y`? The constant test is usually last.

_This could even be a new pylint checker, to detect the reversed order. Tbh. though I have only ever seen the "correct" one `x.y is None`._

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+        # Assume true if uninferable
```
```suggestion
        # Assume true if Uninferable
```

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+        if inferred is util.Uninferable:
+            return True
+
+        return self.negate ^ matches(inferred, nodes.Const(None))
```
Can you add a small comment here?
It's correct but at least I don't see binary XORs all that often in Python. A comment would help to understand it quicker.

Pierre-Sassoulas replied:
It's done in latest versions: ``# Return the XOR of self.negate and matches(inferred, self.CONST_NONE)``

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+        if inferred is util.Uninferable:
+            return True
+
+        return self.negate ^ matches(inferred, nodes.Const(None))
```
```suggestion
        return self.negate ^ matches(inferred, self.CONST_NONE)
```
Or `NoneConstraint.CONST_NONE`

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+        parent = current_node.parent
+        if isinstance(parent, nodes.If):
+            branch, _ = parent.locate_child(current_node)
+            if branch == "body":
+                constraint = match_constraint(expr, parent.test)
+            elif branch == "orelse":
+                constraint = match_constraint(expr, parent.test, invert=True)
+            else:
+                constraint = None
```
```suggestion
            constraint: Constraint | None = None
            if branch == "body":
                constraint = match_constraint(expr, parent.test)
            elif branch == "orelse":
                constraint = match_constraint(expr, parent.test, invert=True)
```

cdce8p commented on astroid/constraint/constraint.py:
```diff
@@ -0,0 +1,138 @@
+    return constraints
+
+
+ALL_CONSTRAINTS = (NoneConstraint,)
+"""All supported constraint types."""
```
This can be a set.

You could also explore using `__init_subclass__` to auto-register all Constraint subclasses with the base class. Although I'm not sure how it interacts with `abc.ABC`.

Pierre-Sassoulas replied:
It's not a frozenset : ``ALL_CONSTRAINT_CLASSES = frozenset((NoneConstraint,))``

cdce8p commented on astroid/bases.py:
```diff
@@ -129,12 +129,15 @@ def _infer_stmts(
     else:
         name = None
         context = InferenceContext()
+        constraints = {}
```
I like to initialize container types unconditionally. That opens the possibility of adding a type hint which will help type checkers infer other variables later. Like `stmt_constraints` with is currently `set[Unknown]`. That should get better with
```py
constraints: dict[str, dict[nodes.If, constraints.Constraint]] = {}
```

cdce8p reviewed (changes requested):
Thanks for your patients @david-yz-liu! It's immediately visible that you've put a lot of work into it. So far I only had time for a surface level view. Never the less, I wanted to give you a first feedback as this is pretty easy to implement and will make the subsequent review a bit easier.

cdce8p commented on astroid/constraint/__init__.py:
```diff
@@ -0,0 +1,10 @@
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Representations of logical constraints on values used during inference.
+"""
+
+from .constraint import get_constraints
+
+__all__ = ("get_constraints",)
```
Just saw the comment from @DanielNoord https://github.com/PyCQA/astroid/pull/1189#discussion_r834192353

IMO it's fine to leave everything in one module for now. If it get's too big, we can always go back and split it but for 140 lines, it just adds unnecessary complexity.

DanielNoord replied:
I recommended this because of the many "moving" commits in astroid that interfere with git blame. It's a little premature, but it would avoid such churn later on.

But if both of you think it should stay in one file, then please go ahead. And sorry for the churn I caused then.

cdce8p replied:
The git blame issue is an inconvenience, true. However we shouldn't overcomplicate things now just in anticipation of what could be. We're far way from any necessary split up, can add at least another 2000 lines 😄

DanielNoord replied:
Then @david-yz-liu please revert the change. As I said, sorry for the inconvenience. Your work on this PR is much appreciated as it is one of the longer request features of `astroid` and `pylint`.

david-yz-liu commented:
Hello! I apologize for the delay again in making these updates (just have had a lot of other things going on). Notably, I've moved the constraint code back into a single file (`constraint.py`) and removed the `__all__` and instead used the underscore prefix where appropriate.

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```
Since this is a `classmethod`, the type of the first argument should be a class type not an instance.
```suggestion
    @classmethod
    @abstractmethod
    def match(
        cls: type[_ConstraintT], node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```

DanielNoord replied:
Changed this to be `Self` which I think makes more sense here, let me know if I need to revert

cdce8p replied:
`Self` would be awesome. Unfortunately it isn't supported by mypy yet.
The PR has already been merged, but we need to wait for the next release (1.0.0).

DanielNoord replied:
Yeah, because it was already merged I thought it would be okay to use. `pyright` understands it and I don't think we will turn on `mypy` in `astroid` before `1.0.0` gets released so this doesn't give us any additional work (instead it prevents some :smile: )

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+
+    CONST_NONE: nodes.Const = nodes.Const(None)
+
+    @classmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```
```suggestion
    @classmethod
    def match(
        cls: type[_ConstraintT], node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```
Same here

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+        """
+
+    @abstractmethod
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
```
```suggestion
    def satisfied_by(self, inferred: InferenceResult) -> bool:
```
See https://github.com/PyCQA/astroid/pull/1189#discussion_r903134872

`InferenceResult` can be imported from `astroid/typing.py`.

Pierre-Sassoulas replied:
This was done at some point.

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+
+        return None
+
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
```
```suggestion
    def satisfied_by(self, inferred: InferenceResult) -> bool:
```

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                _matches(left, node)
+                and _matches(right, cls.CONST_NONE)
+                or _matches(left, cls.CONST_NONE)
```
`cls.CONST_NONE` is a type error. See: https://github.com/PyCQA/astroid/pull/1189#discussion_r903136395

DanielNoord replied:
I think this might be fixed with using `Self`?

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+            if op in {"is", "is not"} and (
+                _matches(left, node)
+                and _matches(right, cls.CONST_NONE)
+                or _matches(left, cls.CONST_NONE)
+                and _matches(right, node)
```
I would suggest we remove the `None is ...` check. That should not be used in code.
See: https://github.com/PyCQA/astroid/pull/1189#discussion_r903138606

DanielNoord replied:
Removed!

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+    return constraints
+
+
+ALL_CONSTRAINT_CLASSES = (NoneConstraint,)
+"""All supported constraint types."""
```
```suggestion
ALL_CONSTRAINT_CLASSES = frozenset((NoneConstraint,))
"""All supported constraint types."""
```
See: https://github.com/PyCQA/astroid/pull/1189#discussion_r903145312

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+"""All supported constraint types."""
+
+
+def _matches(node1: nodes.NodeNG, node2: nodes.NodeNG) -> bool:
```
```suggestion
def _matches(node1: nodes.NodeNG | bases.Proxy, node2: nodes.NodeNG) -> bool:
```
With `InferenceResult`, `bases.Proxy` will also be passed as `node1`.

Pierre-Sassoulas replied:
This has been done.

cdce8p commented on astroid/constraint.py:
```diff
@@ -0,0 +1,135 @@
+    node: _NameNodes, expr: nodes.NodeNG, invert: bool = False
+) -> Constraint | None:
+    """Returns a constraint pattern for node, if one matches."""
+    for constraint_cls in ALL_CONSTRAINT_CLASSES:
+        constraint = constraint_cls.match(node, expr, invert)
+        if constraint:
+            return constraint
```
Is there a case in which multiple constraint classes could match (in the future)? Does it make sense to change it to an Iterator then, i.e. `yield constraint` instead of `return`?

cdce8p commented on astroid/bases.py:
```diff
@@ -146,9 +149,12 @@ def _infer_stmts(
     if context is not None:
         name = context.lookupname
         context = context.clone()
+        constraints = context.constraints.get(name, {})
```
It's better to avoid assigning a different type to an existing variable. That will mess with type checking and make it more difficult to understand. Better to choose a unique name for it.

Pierre-Sassoulas replied:
I think it's fixed in the latest version ``context.constraints`` return a dict.

cdce8p reviewed (commented):
> Hello! I apologize for the delay again in making these updates (just have had a lot of other things going on).

Not necessary 🙂 We all do.

> Notably, I've moved the constraint code back into a single file (constraint.py) and removed the __all__ and instead used the underscore prefix where appropriate.

👍🏻

--
Left some more style comments.

DanielNoord commented:
I handled the review by Marc and incorporate his comments to help move this PR along. I think we're all eager to get this merged 😄

david-yz-liu commented:
Thanks @DanielNoord and @cdce8p! I wasn't sure exactly when I'd have time to look at this again so very much appreciate the updates. 👍 

BTW on the question of whether multiple constraint *classes* may match a given node (in the future), I think... maybe but not necessarily. For example, if we imagine a hypotheical "`AndConstraint`" from

```python
if x is not None and isinstance(x, int):
    ... x ...
```

and so the if node's `test` is an `and` of two expressions, I could see either yielding two distinct constraints ("not none constraint" and "is int instance" constraint), or a single `AndConstraint` whose children are the two constraints. (Personally I'd prefer going with the latter approach.)

DanielNoord commented:
> BTW on the question of whether multiple constraint _classes_ may match a given node (in the future), I think... maybe but not necessarily. For example, if we imagine a hypotheical "`AndConstraint`" from
> 
> ```python
> if x is not None and isinstance(x, int):
>     ... x ...
> ```
> 
> and so the if node's `test` is an `and` of two expressions, I could see either yielding two distinct constraints ("not none constraint" and "is int instance" constraint), or a single `AndConstraint` whose children are the two constraints. (Personally I'd prefer going with the latter approach.)

I think the concept of `AndConstraint` makes sense. This would remove the need for the `yield` behaviour.

I implemented the `yielding` behaviour for now but that could easily be reverted. @cdce8p what do you think?

david-yz-liu commented:
💯 @Pierre-Sassoulas thank you! And thank you all for the efforts in reviewing this and keeping the branch up to date. I'm super excited to see how this evolves!
